demos/https/iot_demo_https_s3_upload_async.c

```diff
@@ -0,0 +1,538 @@
+    connConfig.privateKeyLen = ( ( IotNetworkCredentials_t * ) pNetworkCredentialInfo )->privateKeySize;
+    connConfig.pNetworkInterface = pNetworkInterface;
+
+    asyncInfo.callbacks.writeCallback = _writeCallback;
+    asyncInfo.callbacks.responseCompleteCallback = _responseCompleteCallback;
+    asyncInfo.pPrivData = &uploadSuccess;
+
+    /* The path is everything that is not the address. It also includes the query. So we get the strlen( pPath ) to
+     * acquire everything following in IOT_DEMO_HTTPS_PRESIGNED_PUT_URL. */
+    reqConfig.pPath = pPath;
+    reqConfig.pathLen = strlen( pPath );
+    reqConfig.pHost = pAddress;
+    reqConfig.hostLen = addressLen;
+
+    /* The PUT method is used to upload an object to S3 because it is simpler than using POST. POST requires AWS S3
+     * access specific header fields in the body of the message in a multipart/form-data encoded message. See
+     * https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html for more information about POST object. See
+     * https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html for more information about PUT object. */
+    reqConfig.method = IOT_HTTPS_METHOD_PUT;
+    reqConfig.isNonPersistent = false;
+    reqConfig.userBuffer.pBuffer = _pReqUserBuffer;
+    reqConfig.userBuffer.bufferLen = sizeof( _pReqUserBuffer );
+    reqConfig.isAsync = true;
+    reqConfig.u.pAsyncInfo = &asyncInfo;
+
+    respConfig.userBuffer.pBuffer = _pRespUserBuffer;
+    respConfig.userBuffer.bufferLen = sizeof( _pRespUserBuffer );
+    respConfig.pSyncInfo = NULL;
+
+    /* Create the semaphore for waiting for the whole upload response to finish. */
+    uploadFinishedSemCreated = IotSemaphore_Create( &( _uploadFinishedSem ), 0 /* Initial value. */, 1 /* Max value. */ );
+
+    if( uploadFinishedSemCreated == false )
+    {
+        IotLogError( "Failed to create a semaphore to wait for the response to finish." );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Initialize the HTTPS library. */
+    httpsClientStatus = IotHttpsClient_Init();
+
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "An error occurred initializing the HTTPS library. Error code: %d", httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Initialize the request to initialize internal request context, write the HTTP request first line to the
+     * _pReqUserBuffer header space, and write headers User-Agent and Host. */
+    httpsClientStatus = IotHttpsClient_InitializeRequest( &reqHandle, &reqConfig );
+
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "An error occurred in IotHttpsClient_InitializeRequest() with error code: %d", httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Connect to S3. */
+    for( connIndex = 1; connIndex <= IOT_DEMO_HTTPS_CONNECTION_NUM_RETRY; connIndex++ )
```

Comment:
Index is confusing naming to use here, since I think that would otherwise give the impression of a zero-based count (e.g. for accessing an array).

Reply:
Good point